src/lib/mina/actions/offchain-state.ts

```diff
@@ -190,271 +211,342 @@ function OffchainState<
   const height = logTotalCapacity + 1;
   class IndexedMerkleMapN extends IndexedMerkleMap(height) {}
 
-  // setup internal state of this "class"
-  let internal = {
-    _contract: undefined as OffchainStateContract | undefined,
-    _contractClass: undefined as OffchainStateContractClass | undefined,
-    _merkleMap: undefined as IndexedMerkleMapN | undefined,
-    _valueMap: undefined as Map<bigint, Field[]> | undefined,
-
-    get contract() {
-      return assertDefined(
-        internal._contract,
-        'Must call `setContractInstance()` first'
-      );
-    },
-  };
-  const onchainActionState = async () => {
-    let actionState = (await internal.contract.offchainState.fetch())
-      ?.actionState;
-    assert(actionState !== undefined, 'Could not fetch action state');
-    return actionState;
-  };
-
-  const merkleMaps = async () => {
-    if (internal._merkleMap !== undefined && internal._valueMap !== undefined) {
-      return { merkleMap: internal._merkleMap, valueMap: internal._valueMap };
-    }
-    let actionState = await onchainActionState();
-    let { merkleMap, valueMap } = await fetchMerkleMap(
-      height,
-      internal.contract,
-      actionState
-    );
-    internal._merkleMap = merkleMap;
-    internal._valueMap = valueMap;
-    return { merkleMap, valueMap };
-  };
+  const emptyMerkleMapRoot = new IndexedMerkleMapN().root;
 
   let rollup = OffchainStateRollup({
     logTotalCapacity,
     maxActionsPerProof,
     maxActionsPerUpdate,
   });
 
-  function getContract(): OffchainStateContract {
-    let Contract = assertDefined(
-      internal._contractClass,
-      'Must call `setContractInstance()` or `setContractClass()` first'
-    );
-    return contract(Contract);
-  }
+  function OffchainStateInstance(): OffchainStateInstance<Config> {
+    type InternalState = {
+      _contract: OffchainStateContract<Config> | undefined;
+      _contractClass: OffchainStateContractClass<Config> | undefined;
+      contract: OffchainStateContract<Config>;
+      contractClass: OffchainStateContractClass<Config>;
+      merkleMap: IndexedMerkleMapN;
+      valueMap: Map<bigint, Field[]>;
+    };
 
-  function maybeContract() {
-    try {
-      return getContract();
-    } catch {
-      return internal.contract;
+    type InitializedInternalState = {
```

Comment:
This type seems to be unused. Do you plan to use it?

Reply:
No, it can be removed, I'll do that in my next commit